### ISSUE TITLE
3-fasttext, no bigrams in function predict_sentiment

### ISSUE THREAD

**Issue by 404akhan** [bentrevett/pytorch-sentiment-analysis#26]
Update `predict_sentiment` function at the end to include bigrams.

In addition, since most bigrams do not have pretrained word embeddings they are initialized as all zeros, which is a bad practice. Try to initialize randomly (as in xavier initializer).

**Comment by bentrevett** [comment]
Thanks for spotting these! I'll admit I wasn't even aware of `unk_init` and always assumed TorchText just randomly initialized words outside of the vector vocab, not just set them all to zero.

I'll sort them out now.

**Comment by 404akhan** [comment]
@bentrevett now <pad> and <unk> tokens get non zero weights in embedding. Logically you dont want <pad> or <unk> to influence anything, thus weights of zero are optimal for them, imo. So, that's a new problem :D. Unless, we hope them to converge to zero automatically, not sure. I think fixing <pad> to zeros throughout whole training is very reasonable idea, not sure how to do that though.

**Comment by bentrevett** [comment]
Yep, looks like we have to manually make the unk and padding vectors zero. We can do this with:

```python
model.embedding.weight.data[TEXT.vocab.stoi['<unk>']] = torch.zeros(EMBEDDING_DIM)
model.embedding.weight.data[TEXT.vocab.stoi['<pad>']] = torch.zeros(EMBEDDING_DIM)
```

after we load the pre-trained embeddings. 

We can also fix the padding tokens to stay zeros by passing the index of the padding token (1 by default, or got from `TEXT.vocab.stoi['<pad>']) as the `padding_idx` argument to the `nn.Embedding` layer

```python
self.embedding = nn.Embedding(vocab_size, embedding_dim, padding_idx=padding_idx)
```

**Comment by bentrevett** [comment]
So, PyTorch initializes embedding weights by N(0, 1), so I'm happy to keep `unk_init` to how it is for now.

I might mention how you can define your own initialization if needed and use Xavier as an example as something like:

```
def xavier_init(tensor):
    nn.init.xavier_uniform_(tensor.unsqueeze(0))
    return tensor.squeeze(0)

TEXT.build_vocab(train_data, max_size=25000, vectors="glove.6B.100d", unk_init=xavier_init)
```

**Comment by 404akhan** [comment]
okay, thanks a lot!

**Comment by 404akhan** [comment]
So `fan-in` in our case is 25,000 (vocab size) and `fan-out` 100 (embedding dimension).
I suppose you cannot just set `fan-in` to 1 by unsqueezing. You should manually perform division by vocab size so that it will be xavier. I understand that this is a bit ugly.

**Comment by bentrevett** [comment]
Ah, you're right. That does make it a bit ugly unfortunately. If it comes out on top during the testing then I'll try and figure out a way to implement it without some ugly hack.

**Comment by bentrevett** [comment]
Ran some experiments with this and they basically get the same results for the models used in this tutorial so I'm going to leave it with the Normal initialization for now.
